### PR TITLE
Loki Log Context: Always show label filters with at least one parsed label

### PR DESCRIPTION
**Pull request by svennergr**
**What is this feature?**
Removes a check if a query contains a parser and only rely on the `parsedLabels` field to determine if the second search input box is present.

image

### CODE REVIEW

File: public/app/plugins/datasource/loki/LogContextProvider.ts [hunk 1]
```diff
@@ -248,6 +248,14 @@ export class LogContextProvider {
...
         }
       }
+    } else if (query && isQueryWithParser(query.expr).parserCount === 0) {
+      // this case handles labels added from structured metadata, because the label is `fromParser` but the query does not contain a parser
+      const parsedLabels = contextFilters.filter((filter) => filter.fromParser && filter.enabled);
+      for (const parsedLabel of parsedLabels) {
+        if (parsedLabel.enabled) {
+          expr = addLabelToQuery(expr, parsedLabel.label, '=', parsedLabel.value, LabelType.StructuredMetadata);
+        }
+      }
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/LogContextProvider.ts:258]: I think it make sense, but I am wondering, if we want to refactor this whole logic cause these both `ifs` are very similar, only difference is if we add parser or not.

Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/LogContextProvider.ts:258]: Also, maybe we should change `fromParser` prop to `indexed` and mark indexed labels instead, cause `fromParser` now has both - parsed and structured metadata labels, which is not intuitive. What do you think?

File: public/app/plugins/datasource/loki/components/LokiContextUi.tsx [hunk 2]
```diff
@@ -253,8 +252,8 @@ export function LokiContextUi(props: LokiContextUiProps) {
     };
   }, []);
 
-  // Currently we support adding of parser and showing parsed labels only if there is 1 parser
-  const showParsedLabels = origQuery && isQueryWithParser(origQuery.expr).parserCount === 1 && parsedLabels.length > 0;
+  // If there's any "parsedLabel", that also includes structured metadata, we show the parsed labels input
+  const showParsedLabels = parsedLabels.length > 0;
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/LokiContextUi.tsx:256]:
```suggestion
  const showNonIndexedLabels = nonIndexedLabels.length > 0;
```

Could we change it to something like this - to cover also strucutred metadata?

**Review by ivanahuckova — COMMENTED** [review]
In general make sense and LGTM, but I think we should refactor this a little, to make namings of variables based on loki labels types. And to reduce repetition.

File: public/app/plugins/datasource/loki/LogContextProvider.ts (hunk 1 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/LogContextProvider.ts:258]: > Also, maybe we should change fromParser prop to indexed and mark indexed labels instead, cause fromParser now has both - parsed and structured metadata labels, which is not intuitive. What do you think?

I think that's a good idea, I would still just change it to `nonIndexed` for now. So we don't introduce a logical change that might cause a bug.

File: public/app/plugins/datasource/loki/LogContextProvider.ts [hunk 3]
```diff
@@ -237,15 +237,26 @@ export class LogContextProvider {
...
+      let hasParser = false;
+      if (isQueryWithParser(query.expr).parserCount === 1) {
+        hasParser = true;
+        const parser = getParserFromQuery(query.expr);
+        if (parser) {
+          expr = addParserToQuery(expr, parser);
+        }
+      }
+
+      const parsedLabels = contextFilters.filter((filter) => filter.nonIndexed && filter.enabled);
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/LogContextProvider.ts:250]:
```suggestion
      const nonIndexedLabels = contextFilters.filter((filter) => filter.nonIndexed && filter.enabled);
```

I would rename this one as well.

File: public/app/plugins/datasource/loki/LogContextProvider.ts [hunk 4]
```diff
@@ -368,7 +379,9 @@ export class LogContextProvider {
         return { ...contextFilter };
       });
 
-      const isAtLeastOneRealLabelEnabled = newContextFilters.some(({ enabled, fromParser }) => enabled && !fromParser);
+      const isAtLeastOneRealLabelEnabled = newContextFilters.some(
+        ({ enabled, nonIndexed: fromParser }) => enabled && !fromParser
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/LogContextProvider.ts:383]:
```suggestion
        ({ enabled, nonIndexed }) => enabled && nonIndexed
```

File: public/app/plugins/datasource/loki/components/LokiContextUi.tsx [hunk 5]
```diff
@@ -156,7 +155,7 @@ export function LokiContextUi(props: LokiContextUiProps) {
       return;
     }
 
-    if (contextFilters.filter(({ enabled, fromParser }) => enabled && !fromParser).length === 0) {
+    if (contextFilters.filter(({ enabled, nonIndexed: fromParser }) => enabled && !fromParser).length === 0) {
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/LokiContextUi.tsx:158]:
```suggestion
    if (contextFilters.filter(({ enabled, nonIndexed }) => enabled && nonIndexed).length === 0) {
```

File: public/app/plugins/datasource/loki/components/LokiContextUi.tsx [hunk 6]
```diff
@@ -253,8 +252,8 @@ export function LokiContextUi(props: LokiContextUiProps) {
     };
   }, []);
 
-  // Currently we support adding of parser and showing parsed labels only if there is 1 parser
-  const showParsedLabels = origQuery && isQueryWithParser(origQuery.expr).parserCount === 1 && parsedLabels.length > 0;
+  // If there's any "parsedLabel", that also includes structured metadata, we show the parsed labels input
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/LokiContextUi.tsx:255]:
```suggestion
  // If there's any nonIndexed labels, that includes structured metadata and parsed labels, we show the nonIndexed labels input
```

**Review by ivanahuckova — COMMENTED** [review]
I left couple of comments. There are still places where we call nonIndexed labels parsed. Or when we use parsed, it means parsed and structured metadata. I like that you have came up with property `nonIndexed` and I am wondering, if we could just use it everywhere and stop using `fromParser` if  they are not specifically parsed. I think it will make it more clear.

File: public/app/plugins/datasource/loki/components/LokiContextUi.tsx (hunk 6 above)
Comment (svennergr) [reply public/app/plugins/datasource/loki/components/LokiContextUi.tsx]: Good catches! I renamed everything in c84e359e0ac

**Review by ivanahuckova — APPROVED** [review]
LGTM! Thank you for going beyond and updating variable namings for readability.

**Pull request merged**